pyo3-macros-backend/src/utils.rs

```diff
@@ -166,3 +166,9 @@ pub fn apply_renaming_rule(rule: RenamingRule, name: &str) -> String {
         RenamingRule::Uppercase => name.to_uppercase(),
     }
 }
+
+pub(crate) fn is_abi3() -> bool {
+    cfg!(feature = "abi3")
+        || std::env::var_os("PYO3_USE_ABI3_FORWARD_COMPATIBILITY")
+            .map_or(false, |os_str| os_str == "1")
+}
```

Comment:
It's a bit unfortunate to have this duplication here in the macro code, but unfortunately we don't depend on `pyo3-build-config` since #2243. Given the FIXME I added to `method.rs` I would prefer to not worry about this for now and tidy up in future.

Reply:
But is it really necessary? Couldn't we use [`cargo:rust-cfg`](https://doc.rust-lang.org/cargo/reference/build-scripts.html#rustc-cfg) in the build script to automatically enable the `abi3` feature, e.g. by printing `cargo:rust-cfg=feature=abi3`?

Reply:
I think that's better, yes, as it ensures things are in sync.  I'll change to that, even if we end up reworking this eventually for the FIXME.

Reply:
Ah, actually, it looks like `pyo3-macros-backend` does not currently have a `build.rs`, and it's not clear to me that it'd be correct to try to enable this feature from PyO3's main `build.rs`. Would you be open to just leaving this as it is, rather than introducing a new build script?

Reply:
> it's not clear to me that it'd be correct to try to enable this feature from PyO3's main build.rs.

This is what I would have gone for. The environment variable is basically a way to add the `abi3` to any `Cargo.toml` referencing `pyo3` without being able to edit said `Cargo.toml` files. This would then do anything else that is happening based on enabling that feature, including properly configuring our macro crates. But I'd say that `pyo3`'s self-configuration is the right place for this.

Reply:
Right, but quoting the Cargo docs:

> Note that this does not affect Cargo’s dependency resolution. This cannot be used to enable an optional dependency, or enable other Cargo features.

So I think that adding `cargo:rust-cfg=feature="pyo3-macros-backend/abi3` to the build script won't work as intended?

I think instead what I'll do is push a commit to revert #2243 and pull the configuration from `pyo3-build-config` again, and centralise it there.

Reply:
> So I think that adding cargo:rust-cfg=feature="pyo3-macros-backend/abi3 to the build script won't work as intended?

I don't think the build script of crate X can directly influence the compilation of crate Y even if Y is a dependency of X. My idea here, was to only emit `cargo:rust-cfg=feature=abi` and have the feature forwarding from `Cargo.toml` "just work", but I guess this is my think'o: This is not really enabling the feature from Cargo's point of view, but just passing compiler flags as if the feature were enabled.

> I think instead what I'll do is push a commit to revert https://github.com/PyO3/pyo3/pull/2243 and pull the configuration from pyo3-build-config again, and centralise it there.

Sorry that this became a rabbit hole which you actually anticipated and wanted to avoid. But I also think that spooky-action-at-a-distance via environment variables is [a good thing](https://blog.sunfishcode.online/no-ghosts/) to minimise as much as possible, so I would indeed be glad if the above could be implemented. Thanks!